### PR TITLE
Refactor metadata serializers

### PR DESCRIPTION
**Pull request by tmat**
Removes post-serialization state from MetadataBuilder.

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            // reserved
+            builder.WriteUInt32(0);
+
+            // metadata version length
+            builder.WriteUInt32(MetadataSizes.MetadataVersionPaddedLength);
+
+            int n = Math.Min(MetadataSizes.MetadataVersionPaddedLength, metadataVersion.Length);
+            for (int i = 0; i < n; i++)
+            {
+                builder.WriteByte((byte)metadataVersion[i]);
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.cs]: Should we asserting or validating that metadataVersion is ASCII-safe?

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.cs]: Good catch.

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.cs]: I need to write more tests of the new Builders.

**Comment by tmat** [comment]
@nguerrera Added more tests and validation.

**Comment by tmat** [comment]
@dotnet-bot retest Innerloop Ubuntu14.04 Debug Build and Test please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/ubuntu14.04_debug_prtest/2319/
// Retest reason: https://github.com/dotnet/corefx/issues/9543

**Comment by tmat** [comment]
@dotnet-bot retest Innerloop OSX Release Build and Test please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/osx_release_prtest/2335/
// Retest reason: #9543

**Comment by tmat** [comment]
@dotnet-bot retest Innerloop OSX Release Build and Test please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/osx_release_prtest/2346/
// Retest reason:

**Comment by tmat** [comment]
@nguerrera If you have more feedback I'll follow up. Need to get a new build with these changes to update Roslyn.

**Pull request merged**